docs/topic/cluster-config.rst

```diff
@@ -0,0 +1,150 @@
+
+We request our cluster masters to have `highly available masters <https://cloud.google.com/kubernetes-engine/docs/concepts/regional-clusters>`_
+with ``--region`` parameter. This specifies the region where our 3 master nodes
+will be spread across in different zones. It costs us nothing extra, so we should
+always do it.
+
+By default, asking for highly available masters also asks for 3x the node count,
+spread across multiple zones. We don't want that, since all our user pods have
+in-memory state & can't be relocated. Specifying ``--node-locations`` explicitly
+lets us control how many and which zones the nodes are located in.
+
+Region / Zone selection
+-----------------------
+
+We generally use the ``us-central1`` region and a zone in it for our clusters -
+simply because that is where we have asked for `quota <https://cloud.google.com/compute/quotas>`_.
+
+There are regions closer to us, but latency hasn't really mattered so we are
+currently still in us-central1. There are also unsubstantiated rumors that us-central1 is their
+biggest data center and hence less likely to run out of quota.
+
+Ubuntu operating system
+-----------------------
+
+Since we use :ref:`NFS for user home directories <topic/storage>`, we select
+Ubuntu as our `node operating system <https://cloud.google.com/kubernetes-engine/docs/concepts/node-images>`_.
+The default (Container Optimized OS) does not have NFS support enabled.
+
+Disk Size
+---------
+
+``--disk-size`` sets the size of the root disk on all the kubernetes nodes. This
+isn't used for any persistent storage such as user home directories. It is only
+used ephemerally for the operations of the cluster - primarily storing docker
+images and other temporary storage. We can make this larger if we use a large number
+of big images, or if we want our image pulls to be faster (since disk performance
+`increases with disk size <https://cloud.google.com/compute/docs/disks/performance>`_
+).
+
+``--disk-type=pd-standard`` gives us standard spinning disks, which are cheaper. We
+can also request SSDs instead with ``--disk-type=pd-ssd`` - it is much faster,
+but also much more expensive. 
+
+Node size
+---------
+
+``--machine-type`` lets us select how much `RAM and CPU <https://cloud.google.com/compute/docs/machine-types>`_
+each of our nodes have. For non-trivial hubs, we generally pick ``n1-highmem-8``, with 52G
+of RAM and 8 cores. This is based on the following heuristics:
+
+#. Students generally are memory limited than CPU limited. In fact, while we
+   have a hard limit on memory use per-user pod, we do not have a CPU limit -
+   it hasn't proven necessary.
+
+#. We try overprovision clusters by about 2x - so we try to fit about 100G of total RAM
+   use in a node with about 50G of RAM. This is accomplished by setting the memory
+   request to be about half of the memory limit on user pods. This leads to massive
+   cost savings, and works out ok.
+
+#. There is a kubernetes limit on 100 pods per node.
```

Comment:
I think this is somewhat outdated, and that it is now "about 100" as I recall it is closer to 110.